timemachine/cpp/src/local_md_potentials.cu

```diff
@@ -0,0 +1,190 @@
+    const int seed,
+    const double radius,
+    const double k,
+    const cudaStream_t stream) {
+
+    curandErrchk(curandSetStream(cr_rng_, stream));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+    // Reset the generator offset to ensure same values for the same seed are produced
+    // Simply reseeding does NOT produce identical results
+    curandErrchk(curandSetGeneratorOffset(cr_rng_, 0));
+
+    // Set the array to all N, which indicates to ignore that idx
+    k_initialize_array<unsigned int><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(N_, d_free_idxs_.data, N_);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Generate values between (0, 1.0]
+    curandErrchk(curandGenerateUniform(cr_rng_, probability_buffer_.data, round_up_even(N_)));
+
+    std::mt19937 rng;
+    rng.seed(seed);
+    std::uniform_int_distribution<unsigned int> random_dist(0, local_idxs.size() - 1);
+
+    unsigned int reference_idx = local_idxs[random_dist(rng)];
+
+    const double kBT = BOLTZ * temperature;
+    // Select all of the particles that will be free
+    k_log_probability_selection<float><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(
+        N_, kBT, radius, k, reference_idx, d_x_t, d_box_t, probability_buffer_.data, d_free_idxs_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    this->_setup_free(reference_idx, radius, k, stream);
+}
+
+void LocalMDPotentials::_setup_free(
+    const unsigned int reference_idx, const double radius, const double k, const cudaStream_t stream) {
+    const int tpb = warp_size;
+
+    LessThan select_op(N_);
+
+    // Partition the free idxs into the row idxs
+    gpuErrchk(cub::DevicePartition::If(
+        d_temp_storage_buffer_->data,
+        temp_storage_bytes_,
+        d_free_idxs_.data,
+        d_row_idxs_.data,
+        num_selected_buffer_.data,
+        N_,
+        select_op,
+        stream));
+
+    gpuErrchk(cudaMemcpyAsync(
+        p_num_selected_.data,
+        num_selected_buffer_.data,
+        1 * sizeof(*p_num_selected_.data),
+        cudaMemcpyDeviceToHost,
+        stream));
+    gpuErrchk(cudaStreamSynchronize(stream));
+
+    const int num_row_idxs = p_num_selected_.data[0];
+    const int num_col_idxs = N_ - num_row_idxs;
+
+    if (num_row_idxs == 0) {
+        throw std::runtime_error("LocalMDPotentials setup has no free particles selected");
+    }
+
+    // The reference particle will always be in the column idxs
+    if (num_row_idxs == N_ - 1) {
+        fprintf(stderr, "LocalMDPotentials setup has entire system selected\n");
+    }
+
+    k_construct_bonded_params<<<ceil_divide(num_row_idxs, tpb), tpb, 0, stream>>>(
+        num_row_idxs, N_, reference_idx, k, 0.0, radius, d_row_idxs_.data, restraints_.data, bond_params_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Setup the flat bottom restraints
+    bound_restraint_->set_params_device(std::vector<int>({num_row_idxs, 3}), bond_params_.data, stream);
```

Comment:
Note to self, pinged @badisa offline about this:

I think setting new parameters is fine (overriding existing ones but maintaining shape), but setting new shapes is going to be hell to debug. Not sure why we ever supported this.

Reply:
(maybe @mcwitt has an idea)

Reply:
This was originally done to try and avoid using more memory than we had to. With this approach and the fact that these contexts will be few in number, we can remove it.

We can allocate the potential parameters like you asked here: https://github.com/proteneer/timemachine/pull/1004#discussion_r1169039369 instead. I would be inclined to make this change in a separate PR and combined it with the @mcwitt  suggestion of removing `shape` as a concept from the BoundPotentials, as it always ends up converted to a shape in practice.

Reply:
sgtm!